test/dataframerow.jl

```diff
@@ -3,226 +3,281 @@ module TestDataFrameRow
-@test length(gd[DataFrameRow(df6, 1, :)]) > 0
-@test_throws KeyError gd[DataFrameRow(df6, 2, :)]
-@test isempty(DataFrames.findrows(gd, df6, (gd.df[1],), (df6[1],), 2))
-
-# grouping empty frame
-gd = DataFrames.group_rows(DataFrame(x=Int[]))
-@test length(unique(gd.groups)) == 0
-
-# grouping single row
-gd = DataFrames.group_rows(df5[1:1,:])
-@test length(unique(gd.groups)) == 1
-
-# getproperty, setproperty! and propertynames
-r = DataFrameRow(df, 1, :)
-@test Base.propertynames(r) == names(df)
-@test r.a === 1
-@test r.b === 2.0
-@test copy(r[[:a,:b]]) === (a=1, b=2.0)
-
-r.a = 2
-@test r.a === 2
-r.b = 1
-@test r.b === 1.0
-
-# keys, values and iteration, size
-@test keys(r) == names(df)
-@test values(r) == (df[1, 1], df[1, 2], df[1, 3], df[1, 4])
-@test collect(pairs(r)) == [:a=>df[1, 1], :b=>df[1, 2], :c=>df[1, 3], :d=>df[1, 4]]
-
-@test haskey(r, :a)
-@test haskey(r, 1)
-@test !haskey(r, :zzz)
-@test !haskey(r, 0)
-@test !haskey(r, 1000)
-@test_throws ArgumentError haskey(r, true)
-
-x = DataFrame(ones(5,4))
-dfr = view(x, 2, 2:3)
-@test names(dfr) == names(x)[2:3]
-dfr = view(x, 2, [4,2])
-@test names(dfr) == names(x)[[4,2]]
-
-x = DataFrame(ones(10,10))
-r = x[3, [8, 5, 1, 3]]
-@test length(r) == 4
-@test lastindex(r) == 4
-@test ndims(r) == 1
-@test ndims(typeof(r)) == 1
-@test size(r) == (4,)
-@test size(r, 1) == 4
-@test_throws BoundsError size(r, 2)
-@test keys(r) == [:x8, :x5, :x1, :x3]
-r[:] = 0.0
-r[1:2] = 2.0
-@test values(r) == (2.0, 2.0, 0.0, 0.0)
-@test collect(pairs(r)) == [:x8 => 2.0, :x5 => 2.0, :x1 => 0.0, :x3 => 0.0]
-
-# convert
-df = DataFrame(a=[1, missing], b=[2.0, 3.0])
-dfr = DataFrameRow(df, 1, :)
-@test convert(Vector, dfr)::Vector{Union{Float64, Missing}} == [1.0, 2.0]
-@test convert(Vector{Int}, dfr)::Vector{Int} == [1, 2]
-@test Vector(dfr)::Vector{Union{Float64, Missing}} == [1.0, 2.0]
-@test Vector{Int}(dfr)::Vector{Int} == [1, 2]
-
-# copy
-df = DataFrame(a=Union{Int, Missing}[1, 2, 3, 1, 2, 2],
-               b=[2.0, missing, 1.2, 2.0, missing, missing],
-               c=["A", "B", "C", "A", "B", missing])
-@test copy(DataFrameRow(df, 1, :)) == (a = 1, b = 2.0, c = "A")
-@test isequal(copy(DataFrameRow(df, 2, :)), (a = 2, b = missing, c = "B"))
-
-# parent and parentindices
-@test parent(df[2, :]) === df
-@test parentindices(df[2, :]) == (2, Base.OneTo(3))
-@test parent(df[1, 1:3]) === df
-@test parentindices(df[1, [3,2]]) == (1, [3, 2])
-sdf = view(df, [4,3], [:c, :a])
-@test parent(sdf[2, :]) === df
-@test parentindices(sdf[2, :]) == (3, [3, 1])
-@test parent(sdf[1, 1:2]) === df
-@test parentindices(sdf[1, [2, 2]]) == (4, [1, 1])
-
-# iteration and collect
-ref = ["a", "b", "c"]
-df = DataFrame(permutedims(ref))
-dfr = df[1, :]
-@test Base.IteratorEltype(dfr) == Base.EltypeUnknown()
-@test collect(dfr) == ref
-@test eltype(collect(dfr)) === String
-for (v1, v2) in zip(ref, dfr)
-    @test v1 == v2
+@testset "getindex and setindex!" begin
+    df = DataFrame(a=Union{Int, Missing}[1, 2, 3, 1, 2, 2],
```

Comment:
Maybe we could define this only once at the top of the file, and copy it before modifying it?

Reply:
I will do this. But this was exactly where the mess was `df` was overwritten several times + modified in place and you could not tell what is the actual value of `df` in a given group of tests.

Reply:
actually `deepcopy` is needed